MaxEnt.py

```diff
@@ -1,181 +1,122 @@
-
-            dL_dw += dP_dw
-
-        # Updates the weights.
-        w = w - eta * dL_dw
-
-        # Replaces negative weights with 0.
-        w[w < 0] = 0
-
-        # Store history.
-        history.append(w.squeeze())
-
-    return w, history
-
-
-def sorted_learned_weights(constraint, weight) -> dict[str, float]:
-    learned = {}
-    for cn, cw in zip(constraint, weight.squeeze()):
-        learned[cn] = np.round(cw, 3)
-
-    learned = sorted(learned.items(), key=lambda x: x[1], reverse=True)
+    Methods:
+        sorted_by_weights: Returns sorted (constraint, weight) tuples.
+        compute_probabilities: Returns the predicted probability distribution for any given violation profile.
+    """
 
-    for pair in learned:
-        print(f"{pair[0]:<20}: {pair[1]:<10}")
+    def __init__(self, constraints, weights):
+        self.const = constraints
+        self.w = weights
 
-    return dict(learned)
+    def sorted_by_weights(self):
+        """Displays constraints from highest to lowest weights.
 
+        Returns:
+            list[tuple[str, float]]: a list of (constraint, weight) tuple sorted from highest to lowest weighted.
+        """
+        name_weight_pair = {}
+        for cn, cw in zip(self.const, self.w.squeeze()):
+            name_weight_pair[cn] = np.round(cw, 3)
 
-def predict_prob(vio, cand):
-    with open("predictions.csv", "w", newline="") as file:
-        writer = csv.writer(file)
-        writer.writerow(["UR/Candidate", "Probability"])
-        for t in range(len(vio)):
-            P = compute_probabilities(vio[t], w).squeeze()
-            writer.writerow([urs[t]])
+        sorted_pairs = sorted(
+            name_weight_pair.items(), key=lambda x: x[1], reverse=True
+        )
 
-            for c in range(len(cand[t])):
-                writer.writerow([cand[t][c], f"{round(P[c]*100, 1)}%"])
+        return sorted_pairs
+
+    def compute_probabilities(self, violations):
+        """Predicts probabilities for a particular violation profile.
+
+        Args:
+            violations (np.array[np.array[int]]): the violation profiles of a list of tableaux.
+
+        Returns:
+            np.array[np.array[float]]: predicts probability for each candidate of given URs.
+        """
+        harmonies = np.dot(violations, self.w)
+        exp = np.exp(-harmonies)
+        marginalization = np.sum(exp, axis=0)
+        return exp / marginalization
+
+    def SGD_learn(self, candidates, violations, srs, *regs, iters=10000, eta=0.05):
+        """Updates constraint weights using SGD.
+
+        Args:
+            candidates (np.array[np.array[str]]): candidates considered for each UR.
+            violations (np.array[np.array[int]]): the violation profiles of a list of tableaux.
+            srs (np.array[np.array[float]]): probability of a surface form observed for a given UR.
+            *regs: a variable number of regularization terms. Can be None.
```

Comment:
Just wanted to double-check that if we were to call this function:

```
m = MaxEnt(...)
m.SGD_learn(cand, vios, obs, 10000, 0.5)
```

it wouldn't assign anything to ```regs```? Or if we were to do:

```
m = MaxEnt(...)
m.SGD_learn(cand, vios, obs, iters=10000, eta=0.5)
```

```reg``` will default to ```None```?